crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,127 @@
+/// Sender to notify if there is a change in OS interface order
+pub static PATH_CHANGE_BROADCAST: Lazy<Sender<()>> = Lazy::new(|| Sender::new(2));
+/// Vector containing all local interfaces
+pub static LOCAL_ADDRS_CACHE: Lazy<Arc<StdMutex<Vec<if_addrs::Interface>>>> =
+    Lazy::new(|| Arc::new(StdMutex::new(Vec::new())));
+#[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos"))]
+static NETWORK_PATH_MONITOR_START: std::sync::Once = std::sync::Once::new();
+
+#[derive(Debug)]
+/// Struct to monitor network
+pub struct NetworkMonitor {
+    nw_path_monitor_monitor_handle: Option<JoinHandle<io::Result<()>>>,
+    // Mac and Windows use callbacks, whereas in Linux a socket is registered
+    // with netlink multicast group messages
+    #[cfg(target_os = "linux")]
+    linux_monitor_handle: Option<JoinHandle<io::Result<()>>>,
+    #[cfg(target_os = "windows")]
+    windows_monitor_handle: crate::windows::SafeHandle,
+}
+
+impl NetworkMonitor {
+    /// Sets up and spawns network monitor
+    pub async fn new<G>(if_addr: G) -> Result<NetworkMonitor, GetIFError>
+    where
+        G: GetIfAddrs + Clone,
+    {
+        if let Ok(mut guard) = LOCAL_ADDRS_CACHE.lock() {
+            *guard = local_interfaces::gather_local_interfaces(&if_addr)?;
+            telio_log_debug!("local cache {:?}", *guard);
+        }
+
+        let nw_path_monitor_monitor_handle = Some(tokio::spawn({
+            let mut notify = PATH_CHANGE_BROADCAST.subscribe();
+            async move {
+                loop {
+                    match notify.recv().await {
+                        Ok(()) => match local_interfaces::gather_local_interfaces(&if_addr) {
+                            Ok(v) => {
+                                if let Ok(mut guard) = LOCAL_ADDRS_CACHE.lock() {
+                                    telio_log_debug!("Updating local addr cache");
+                                    *guard = v;
+                                }
+                            }
+                            Err(e) => telio_log_warn!("Unable to get local interfaces {e}"),
+                        },
+                        Err(e) => {
+                            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
+                        }
+                    }
+                }
+            }
+        }));
+
+        #[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos"))]
+        NETWORK_PATH_MONITOR_START.call_once(crate::mac::setup_network_monitor);
+
+        #[cfg(target_os = "linux")]
+        let linux_monitor_handle = crate::linux::setup_network_monitor().await;
+
+        #[cfg(target_os = "windows")]
+        let windows_monitor_handle = crate::windows::setup_network_monitor();
+
+        Ok(Self {
+            nw_path_monitor_monitor_handle,
+            #[cfg(target_os = "linux")]
+            linux_monitor_handle,
+            #[cfg(target_os = "windows")]
+            windows_monitor_handle,
+        })
```

Comment:
I think we don't need to have different names for the `monitor_handle`
```suggestion
        #[cfg(target_os = "linux")]
        let monitor_handle = crate::linux::setup_network_monitor().await;

        #[cfg(target_os = "windows")]
        let monitor_handle = crate::windows::setup_network_monitor();

        Ok(Self {
            nw_path_monitor_monitor_handle,
            #[cfg(any(target_os = "windows", target_os = "linux"))]
            monitor_handle,
        })
```